File: examples/python/bad_tests_sample.py
"""
Examples of false-positive test patterns in Python/pytest.
Each test passes green but does not verify correct behavior.
These illustrate the semantic cases (10/11/12/15/18) that the
falsegreen scanner flags for LLM review, plus common weak patterns.
"""
from unittest.mock import patch, MagicMock
import pytest


# ─── Case 10: Mocks the unit under test ────────────────────────────────────

@patch("mymodule.add")
def test_add_case10(mock_add):
    """BAD: patches the SUT itself, then asserts the mock's return value."""
    mock_add.return_value = 5
    result = mock_add(2, 3)     # calling the mock, not the real function
    assert result == 5          # C10 — tests the mock config


@patch("mymodule.db.fetch")
def test_get_user_clean(mock_fetch):
    """CLEAN: patches an edge (db), tests the real function."""
    from mymodule import get_user
    mock_fetch.return_value = {"id": 1, "name": "Alice"}
    user = get_user(1)          # real function under test
    assert user["name"] == "Alice"


# ─── Case 11: Asserts the value fed to the mock ────────────────────────────

def test_price_case11():
    """BAD: stubs price to 100, asserts result equals 100.
    If get_price() just returns mock.price, this is an echo."""
    mock_product = MagicMock()
    mock_product.price = 100
    from mymodule import get_price
    assert get_price(mock_product) == 100   # C11 if get_price returns product.price


def test_price_with_tax_clean():
    """CLEAN: stubs base price, asserts the tax was applied."""
    mock_product = MagicMock()
    mock_product.price = 100
    from mymodule import get_price_with_tax
    assert get_price_with_tax(mock_product) == 110  # 100 + 10% = 110 (from spec)


# ─── Case 12: Re-implements the production formula ─────────────────────────

def test_total_case12():
    """BAD: expected re-implements the formula — both agree on the same
    wrong number if the formula has a bug."""
    price, tax_rate = 100, 0.1
    expected = price + price * tax_rate      # same as calculate_total
    from mymodule import calculate_total
    assert calculate_total(price, tax_rate) == expected  # C12


def test_total_clean():
    """CLEAN: expected comes from the spec, not from a formula copy."""
    from mymodule import calculate_total
    assert calculate_total(100, 0.1) == 110.0    # spec: 100 + 10% = 110


# ─── Case 15: Passes only if another test ran first ────────────────────────

_shared_cache: dict = {}

def test_populate_cache():
    """BAD part 1: populates shared state."""
    _shared_cache["key"] = "value"


def test_read_cache():
    """BAD part 2: C15 — passes only after test_populate_cache ran first."""
    assert _shared_cache["key"] == "value"


# ─── Case 18: Expected contradicts the spec ────────────────────────────────

def test_apply_discount_case18():
    """BAD: the spec says apply_discount(200, 0.15) must return 170.
    This test asserts 200 — no discount was applied.
    If the function is buggy and returns 200, this test passes and freezes
    the bug as 'correct behavior'.
    Oracle: docstring says 'returns price minus (price * rate)'."""
    from mymodule import apply_discount
    assert apply_discount(200, 0.15) == 200  # C18 — asserts the bug


def test_apply_discount_clean():
    """CLEAN: expected from the spec (price - price * rate = 170)."""
    from mymodule import apply_discount
    assert apply_discount(200, 0.15) == 170.0

File: examples/javascript/bad_tests_sample.js
/**
 * False-positive test patterns in JavaScript/Jest.
 *
 * Covers: C10 (mock hoisting, ESM dynamic import), J1 (done() in finally,
 * floating promise), C6 (weak assertion on dynamic property), C27 (swallowed
 * exception via try/catch).
 *
 * Every BAD test passes green but does not verify correct behavior.
 * CLEAN alternatives follow each bad pattern.
 *
 * Function names (fetchData, processOrder, db, getUser, parse, buildResponse)
 * are placeholders -- assume they are imported or defined elsewhere.
 * Patterns apply to Jest (noted where Mocha/Chai differs).
 */


// ─── C10: CJS jest.mock hoisting trap ────────────────────────────────────────

// BAD: jest.mock is hoisted above the require(), so the require gets the mock.
// processOrder is already jest.fn() when the module loads -- the real function
// never runs. The test only exercises the mock configuration, not the SUT.
const { processOrder } = require('./orders')
jest.mock('./orders') // hoisted to top of file by babel-jest / jest transform

test('processOrder returns confirmation', () => {
  processOrder.mockReturnValue({ ok: true })
  expect(processOrder({ id: 1 })).toEqual({ ok: true }) // C10 -- tests the mock config
})

// CLEAN: mock a dependency of orders (the db layer), not orders itself.
// The real processOrder implementation runs; only db.save is replaced.
const { db } = require('./db')
jest.mock('./db')

test('processOrder saves to db', async () => {
  db.save.mockResolvedValue({ id: 42 })
  const { processOrder: realProcessOrder } = jest.requireActual('./orders')
  const result = await realProcessOrder({ item: 'widget' })
  expect(result.id).toBe(42) // asserts on a derived result, not on the mock value
})


// ─── C10: ESM dynamic import mock (Jest ESM mode) ────────────────────────────

// BAD: jest.unstable_mockModule replaces the module under test, then the test
// imports and calls it directly. All assertions land on the mock, not on any
// real logic. Works the same as the CJS hoisting trap -- just ESM syntax.
//
// (Mocha/Chai: same risk with esmock or proxyquire replacing the SUT.)

describe('C10 -- ESM mock', () => {
  beforeAll(() => {
    jest.unstable_mockModule('./formatter', () => ({
      formatDate: jest.fn(() => '2024-01-01'), // BAD: mocking the SUT
    }))
  })

  test('formatDate returns formatted string', async () => {
    const { formatDate } = await import('./formatter') // gets the mock
    expect(formatDate(new Date())).toBe('2024-01-01') // C10 -- tests mock config
  })
})

// CLEAN: mock the dependency (the locale adapter), import the real formatter.
describe('C10 -- ESM mock clean', () => {
  beforeAll(() => {
    jest.unstable_mockModule('./localeAdapter', () => ({
      getLocale: jest.fn(() => 'en-US'), // mocking a dependency, not the SUT
    }))
  })

  test('formatDate uses locale from adapter', async () => {
    const { formatDate } = await import('./formatter') // real implementation
    const result = formatDate(new Date('2024-01-01T00:00:00Z'))
    expect(result).toMatch(/2024/) // asserts on behavior, not on mock value
  })
})


// ─── J1: done() in finally (callback-based async) ────────────────────────────

// BAD: done() fires in the finally block, so it runs even when expect() throws.
// The test exits green whether or not the assertion passed.
//
// (Mocha/Chai: same trap -- done() in finally always resolves the test.)
test('async callback delivers data', (done) => {
  fetchData((err, data) => {
    try {
      expect(data.status).toBe('ok')
    } finally {
      done() // J1 -- done fires even if expect threw; assertion failure is silenced
    }
  })
})

// CLEAN (option A): return a Promise -- no done() needed.
// Jest/Mocha both treat a returned Promise as the test boundary.
test('async callback delivers data -- promise clean', () => {
  return new Promise((resolve, reject) => {
    fetchData((err, data) => {
      if (err) return reject(err)
      try {
        expect(data.status).toBe('ok')
        resolve()
      } catch (e) {
        reject(e) // assertion error propagates -- test fails correctly
      }
    })
  })
})

// CLEAN (option B): done only in the success branch; done.fail in catch.
test('async callback delivers data -- done.fail clean', (done) => {
  fetchData((err, data) => {
    if (err) return done.fail(err)
    try {
      expect(data.status).toBe('ok')
      done()
    } catch (e) {
      done.fail(e) // not done() -- prevents silent green on assertion error
    }
  })
})


// ─── J1: floating promise (unhandled rejection) ───────────────────────────────

// BAD: db.save is async but is not awaited and not returned.
// The test function returns undefined synchronously, Jest marks it green,
// and the rejection (if any) fires after the test boundary.
//
// (Mocha/Chai without --allow-uncaught: same behavior -- test exits before reject.)
const record = { id: 1, name: 'widget' }

test('saves the record', () => {
  db.save(record) // J1 -- async, not awaited; rejection is unhandled
  expect(db.records).toHaveLength(1) // checks state before async write completes
})

// CLEAN: await the async call so the test body suspends until it settles.
test('saves the record -- await clean', async () => {
  await db.save(record)
  expect(db.records).toHaveLength(1) // state is settled before assertion
})

// CLEAN: returning the promise also works (no async/await required).
test('saves the record -- return clean', () => {
  return db.save(record).then(() => {
    expect(db.records).toHaveLength(1)
  })
})


// ─── C6: dynamic property access with weak assertion ─────────────────────────

// BAD: toBeTruthy() on a dynamic property passes for 'admin', 'guest',
// the string 'undefined', 0, or anything except null/false/empty-string/''.
// The bracket notation alone does not indicate a C6 smell -- the weak check does.
test('user has a role', () => {
  const user = getUser(1)
  expect(user['role']).toBeTruthy() // C6 -- passes for any truthy value, including wrong ones
})

// CLEAN: assert the exact expected value.
test('user has the correct role', () => {
  const user = getUser(1)
  expect(user.role).toBe('admin')
})

// CLEAN (not C6): toHaveProperty checks that the key exists on the object.
// Checking key existence is a meaningful assertion, not a weak one.
test('response has required fields', () => {
  const response = buildResponse()
  expect(response).toHaveProperty('status') // existence check -- not weak
  expect(response).toHaveProperty('data')
})

// CLEAN (not C6): toBeDefined() is meaningful when the contract says
// "this field must be present" without prescribing its value.
test('user object has a name field', () => {
  const user = getUser(1)
  expect(user.name).toBeDefined()
  expect(typeof user.name).toBe('string')
})


// ─── C27: try/catch/pass instead of expect().toThrow() ───────────────────────

// BAD: if parse(null) does NOT throw, execution skips the catch block and the
// test exits green with zero assertions. Both paths (throws / does not throw)
// leave the test green.
//
// (Mocha/Chai: same trap -- empty catch swallows the assertion failure.)
test('throws on invalid input', () => {
  try {
    parse(null)
  } catch (e) {
    // C27 -- swallowed. If parse(null) doesn't throw, test still passes.
  }
})

// BAD: the catch asserts on the message, but if parse(null) does not throw,
// no assertion runs at all. C2 (no assertion) and C27 at the same time.
test('error message is correct', () => {
  try {
    parse(null)
    // if parse does not throw, execution falls through here -- no assertion runs
  } catch (e) {
    expect(e.message).toBe('input required')
  }
})

// CLEAN: Jest/Vitest idiomatic -- expect().toThrow() handles both the throw
// check and the message check in one readable assertion.
test('throws on invalid input -- clean', () => {
  expect(() => parse(null)).toThrow('input required')
})

// CLEAN: async variant -- rejectsWith for async functions that reject.
test('async parse rejects on invalid input -- clean', async () => {
  await expect(parseAsync(null)).rejects.toThrow('input required')
})

// CLEAN: if you need the error object for further inspection, use a manual
// try/catch but always add a fail-safe assertion before the try block.
test('throws ValidationError with code -- clean', () => {
  expect.assertions(2) // guards against the empty-catch path
  try {
    parse(null)
  } catch (e) {
    expect(e).toBeInstanceOf(ValidationError)
    expect(e.code).toBe('ERR_NULL_INPUT')
  }
})


// ─── Placeholder stubs (not imported -- defined inline for IDE clarity) ───────

function fetchData(cb) { cb(null, { status: 'ok' }) }
function processOrder() {}
function getUser(id) { return { id, role: 'admin', name: 'Alice' } }
function buildResponse() { return { status: 200, data: {} } }
function parse(input) { if (input === null) throw new Error('input required') }
async function parseAsync(input) { if (input === null) throw new Error('input required') }
class ValidationError extends Error { constructor(msg, code) { super(msg); this.code = code } }
